### PR TITLE
fix for multi file downloads 

### PR DESCRIPTION
**Pull request by luciipak**

### CODE REVIEW

**Comment by jperson** [comment]
we might want to change this to do a chunked streaming response and to write the zip file to disk instead of a io.BytesIO object so the server doesn't use up too much memory when building the .zip file,

that can be done by changing the code to something like

**Comment by jperson** [comment]
last comment, I did a little digging on streaming files from s3 without writing to disk, it looks like it's possible using the `Object.get()['Body']` on the `s3` class

```python
keys = [
    'analyzed/60e88e2a-b101-49e2-9734-96f299fe8959/d5b6b8c8-3671-47d9-a035-884930ae8b97/ML2022136002_VS long_2022_06_14_184951_full.xlsx',
    'analyzed/60e88e2a-b101-49e2-9734-96f299fe8959/d5b6b8c8-3671-47d9-a035-884930ae8b97/ML2022010101__2022_06_13_012109_0.0-30.0.xlsx',
    'analyzed/60e88e2a-b101-49e2-9734-96f299fe8959/d5b6b8c8-3671-47d9-a035-884930ae8b97/ML2022010101__2022_06_13_012109_full.xlsx',
    'analyzed/60e88e2a-b101-49e2-9734-96f299fe8959/d5b6b8c8-3671-47d9-a035-884930ae8b97/ML2022123456_1Hz 35mA_2022_06_24_185409_full.xlsx',
    'analyzed/60e88e2a-b101-49e2-9734-96f299fe8959/d5b6b8c8-3671-47d9-a035-884930ae8b97/ML2022123456_1Hz_50mA_6ms_2022_06_24_185847_full.xlsx',
    'analyzed/60e88e2a-b101-49e2-9734-96f299fe8959/d5b6b8c8-3671-47d9-a035-884930ae8b97/ML2022123456_50mA 1Hz_2022_06_24_185613_full.xlsx',
]

@app.post("/download3")
async def dl3():
    def unzipped_files():
        s3 = boto3.session.Session().resource('s3')
        for k in keys:
            obj = s3.Object(bucket_name='test-pulse3d-uploads', key=k)
            yield k.split('/')[-1], datetime.now(), 0o600, ZIP_64, obj.get()['Body']

    return StreamingResponse(content=stream_zip(unzipped_files()), media_type="application/zip")
```

File: deployments/pulse3d/services/pulse3d/src/main.py [hunk 1]
```diff
@@ -311,3 +318,64 @@ async def soft_delete_jobs(
...
+    except Exception as e:
+        logger.error(f"Failed to download analyses: {repr(e)}")
+        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+
+def _yield_s3_objects(bucket: str, keys: List[str], filenames: List[str]):
+    try:
+        s3 = boto3.session.Session().resource("s3")
+        for (idx, key) in enumerate(keys):
+            obj = s3.Object(bucket_name="test-pulse3d-uploads", key=key)
```
Comment (jperson) [inline deployments/pulse3d/services/pulse3d/src/main.py:377]: you'll want to grab the bucket name from the env so this will work on test and prod

Comment (luciipak) [reply deployments/pulse3d/services/pulse3d/src/main.py:377]: Good catch, just checking  in test if the FE streamsaver package will work with statically generated pages

Comment (luciipak) [reply deployments/pulse3d/services/pulse3d/src/main.py]: time has dropped from 1.5 minutes to .4 min

Comment (luciipak) [reply deployments/pulse3d/services/pulse3d/src/main.py]: I get this build error: 
`Collecting page data ...streamsaver is meant to run on browsers main thread`
```
> Build error occurred
ReferenceError: document is not defined
    at /Users/lucipak/Documents/work/CuriBio/k8s_curi/frontend/pulse3d/src/node_modules/streamsaver/StreamSaver.js:24:66
    at /Users/lucipak/Documents/work/CuriBio/k8s_curi/frontend/pulse3d/src/node_modules/streamsaver/StreamSaver.js:7:24
    at Object.<anonymous> (/Users/lucipak/Documents/work/CuriBio/k8s_curi/frontend/pulse3d/src/node_modules/streamsaver/StreamSaver.js:11:3)
```

might try adding to the service worker?

**Review by jperson — APPROVED** [review]

**Pull request merged**